paper/paper.tex

```diff
@@ -0,0 +1,249 @@
+Note that the time complexity of the sum of $n$ \lstinline|Term| is now $\Theta(n)$.
+
+Julia implements a specialized method for computing the sum of \lstinline|BigInt|s that uses \lstinline|Base.GMP.MPZ.add!|.
+Similarly, before its version v0.21, JuMP used to implement a specialized method for the sum of JuMP expressions.
+The advantage of having a standardized API for mutable addition is that
+only one implementation of \lstinline|sum| is needed.
+This approach of API based on a function that may mutate its first argument in order to allow the same code to work both for mutable and non-mutable type is
+used by the \lstinline|!!| convention in BangBang~\cite{takafumi2021bangbang},
+the mutable API in AbstractAlgebra~\cite{AbstractAlgebra.jl-2017}
+as well as the \lstinline|destructive_add!| function in JuMP v0.20.
+
+\subsection{Should mutate}
+When writing a code that should mutate the first argument of an operation,
+such API that may silently return the result without modifying the first argument is not appropriate.
+For this reason, there are also mutable operations that always mutate the first argument.
+This is the approach used by the \lstinline|!| convention in Julia
+as well as the \lstinline|add_to_expression!| in JuMP.
+
+\subsection{Can mutate ?}
+A third useful feature for users of a mutable API is the ability to
+determine whether objects of a given type can be mutated\footnote{\label{foot:mutate}In this paper, the terminology ``mutate $x$ to $y$'' means mutating $x$ in such a way that its value after the mutation is equal to $y$} to the result of a mutable operation.
+To motivate this, consider the rational Julia type:
+\begin{lstlisting}[language = Julia]
+struct Rational{T}
+    num::T
+    den::T
+end
+\end{lstlisting}
+An implementation of a multiplication function (ignoring the simplification with \lstinline|gcd| for simplicity) that may
+mutate its first argument, say \lstinline|mul!!| to follow BangBang's convention,
+could be:
+\begin{lstlisting}[language = Julia]
+function mul!!(a::Rational{S}, b::Rational{T})
+    if # S can be mutated to `*(::S, ::T)`
+        mul!(a.num, b.num)
+        mul!(a.den, b.den)
+        return a
+    else
+        return a * b
+    end
+end
+\end{lstlisting}
+Note \lstinline|a.num = mul!!(a.num, b.num)| is not an option since
+the \lstinline|Rational| struct is not mutable.
+This third feature would be needed to implement this \lstinline|if| clause.
+
+\subsection{Promotion}
+Algorithms that can exploit mutability often start by creating an accumulator of
+an appropriate type.
+In the example above, \lstinline|sum| start by create the accumulator \lstinline|acc|.
+Another example is matrix multiplication which start by allocating the resulting matrix and then mutate\cref{foot:mutate} it to the result with \lstinline|LinearAlgebra.mul!|.
+To determine the element type of this resulting array,
+LinearAlgebra uses \lstinline|Base.promote_op| which relies on Julia inference.
+
+\section{API}
```

Comment:
This API section is pretty confusing. What's the scope/audience of JuliaCon papers? It's kinda in-the-weeds but without the space necessary to actually elaborate on a lot of points.

Reply:
I'd say package developer wondering if MA would fit their need so I expect a knowledgeable audience